### PR TITLE
fix(select-query-parser): select left join and array instead of objects result types

### PR DESCRIPTION
**Pull request by avallete**
## What kind of change does this PR introduce?
This PR improves the TypeScript typings generated by the SQL query parser, fixing issues with inferred types for certain `!left` joins and relationship queries (e.g., `created_by:users(*)`).

## What is the current behavior?
Currently, for certain queries, the inferred types for relations are incorrect. For instance, the type may be inferred as `{}[]` when the runtime result is a single object `{}`.

**Related issues**
https://github.com/supabase/postgrest-js/issues/546
https://github.com/supabase/supabase/issues/29086

## What is the new behavior?
This PR updates the select-query-parser to handle these cases more accurately by:

- Extending the handling logic for the `!left` operator, similar to the existing logic for the `!inner` operator.
- Returning a single object for relationships when the runtime result is also a single object.

## Additional context
The TypeScript code in select-query-parser is quite complex. I would appreciate a thorough review of this fix.
Additional tests have been added to cover both runtime and typing scenarios. This helped uncover more edge cases, some of which still need further solutions.

### Known Issues:
Currently, there is no distinction between possibly nullable and non-nullable results for zero-to-one or one-to-one relations. To handle this properly, the PR always returns null as a possible option. However, a more accurate solution would involve enhancements like adding an `isOneToOne` option in Relationships that marks foreign keys as nullable or non-nullable based on `columns[]` type definitions.

### PS
Also took the chance to fix the CI workflow by replacing the [deprecated docker-compose](https://github.blog/changelog/2024-04-10-github-hosted-runner-images-deprecation-notice-docker-compose-v1/) with `docker compose` command.

### CODE REVIEW

File: test/db/01-dummy-data.sql [hunk 1]
```diff
@@ -30,3 +30,18 @@ VALUES
...
+
+INSERT INTO user_profiles (id, username)
+VALUES
+  (1, 'supabot'),
+  (2, NULL);
+
+INSERT INTO best_friends(id, first_user, second_user, third_wheel)
+VALUES
+  (1, 'supabot', 'kiwicopple', 'awailas'),
+  (2, 'supabot', 'awailas', NULL);
```
Comment (soedirgo) [inline test/db/01-dummy-data.sql:47]:
```suggestion
  (2, 'supabot', 'awailas', NULL);

```

**Comment by soedirgo** [comment]
Thanks! I think this fixes https://github.com/supabase/supabase/issues/29086 but not https://github.com/supabase/postgrest-js/issues/546 - will investigate that separately

**Review by soedirgo — APPROVED** [review]
